evm/vm/forks/sharding/vmc_handler.py

```diff
@@ -1,34 +1,145 @@
 import logging
 
+import rlp
+
 from web3.contract import (
     Contract,
 )
 
 from eth_utils import (
+    event_signature_to_log_topic,
     is_canonical_address,
     to_checksum_address,
     to_dict,
+    to_tuple,
+)
+
+from evm.rlp.sedes import (
+    address,
+    hash32,
 )
 
 from evm.utils.hexadecimal import (
+    encode_hex,
     decode_hex,
 )
+from evm.utils.numeric import (
+    big_endian_to_int,
+)
 
 from evm.vm.forks.sharding.config import (
     get_sharding_config,
 )
 
 
+class NextLogUnavailable(Exception):
+    pass
+
+
+@to_dict
+def parse_collation_added_data(data_hex):
+    data_bytes = decode_hex(data_hex)
+    score = big_endian_to_int(data_bytes[-32:])
+    is_new_head = bool(big_endian_to_int(data_bytes[-64:-32]))
+    header_bytes = data_bytes[:-64]
+    # [num, num, bytes32, bytes32, bytes32, address, bytes32, bytes32, num, bytes]
+    sedes = rlp.sedes.List([
+        rlp.sedes.big_endian_int,
+        rlp.sedes.big_endian_int,
+        hash32,
+        hash32,
+        hash32,
+        address,
+        hash32,
+        hash32,
+        rlp.sedes.big_endian_int,
+        rlp.sedes.binary,
+    ])
+    header_values = rlp.decode(header_bytes, sedes=sedes)
+    yield 'header', header_values
+    yield 'is_new_head', is_new_head
+    yield 'score', score
+
+
 class VMC(Contract):
 
-    logger = logging.getLogger("evm.chain.sharding.mainchain_handler.VMC")
+    logger = logging.getLogger("evm.chain.sharding.VMC")
+
+    # For handling logs filtering
+    # Event:
+    #   CollationAdded(indexed uint256 shard, bytes collationHeader, bool isNewHead, uint256 score)
+    COLLATION_ADDED_TOPIC = event_signature_to_log_topic(
+        "CollationAdded(int128,bytes4096,bool,int128)"
+    )
 
-    def __init__(self, *args, default_privkey, **kwargs):
+    shards = set()
```

Comment:
You shouldn't use mutable objects as class properties.  This should be set in the constructor as otherwise, all instances of this class will share this variable meaning if it's changed in one, it'll be changed in all of them.

Reply:
Sorry it's my fault. Thank you for pointing out. Will change them.